x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/sender/HttpRequestExecutorService.java

```diff
@@ -0,0 +1,168 @@
+        this(contextHolder, serviceName, null);
+    }
+
+    @SuppressForbidden(reason = "properly rethrowing errors, see EsExecutors.rethrowErrors")
+    public HttpRequestExecutorService(ThreadContext contextHolder, String serviceName, @Nullable Integer capacity) {
+        this.contextHolder = Objects.requireNonNull(contextHolder);
+        this.serviceName = Objects.requireNonNull(serviceName);
+        this.httpContext = HttpClientContext.create();
+
+        if (capacity == null) {
+            this.queue = new LinkedBlockingQueue<>();
+        } else {
+            this.queue = new LinkedBlockingQueue<>(capacity);
+        }
+    }
+
+    /**
+     * Begin servicing tasks.
+     */
+    public void start() {
+        try {
+            while (running.get()) {
+                HttpTask task = queue.take();
+                if (task.shouldShutdown() || running.get() == false) {
+                    running.set(false);
+                    logger.debug(() -> format("Http executor service [%s] exiting", serviceName));
+                } else {
+                    executeTask(task);
+                }
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        } finally {
+            terminationLatch.countDown();
+        }
+    }
+
+    private void executeTask(HttpTask task) {
+        try {
+            contextHolder.preserveContext(task).run();
+        } catch (Exception e) {
+            logger.error(format("Http executor service [%s] failed to execute request [%s]", serviceName, task), e);
+        }
+        EsExecutors.rethrowErrors(ThreadContext.unwrap(task));
+    }
+
+    public int queueSize() {
+        return queue.size();
+    }
+
+    @Override
+    public void shutdown() {
+        if (running.compareAndSet(true, false)) {
+            // if this fails because the queue is full, that's ok, we just want to ensure that queue.take() returns
+            queue.offer(new ShutdownTask());
+        }
+    }
+
+    @Override
+    public List<Runnable> shutdownNow() {
+        shutdown();
+        return new ArrayList<>(queue);
+    }
+
+    @Override
+    public boolean isShutdown() {
+        return running.get() == false;
+    }
+
+    @Override
+    public boolean isTerminated() {
+        return terminationLatch.getCount() == 0;
+    }
+
+    @Override
+    public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+        return terminationLatch.await(timeout, unit);
+    }
+
+    /**
+     * Execute the task at some point in the future.
+     * @param command the runnable task, must be a class that extends {@link HttpTask}
+     */
+    @Override
+    public void execute(Runnable command) {
+        if (command == null) {
+            return;
+        }
+
+        assert command instanceof HttpTask;
+        HttpTask task = (HttpTask) command;
+        task.setContext(httpContext);
```

Comment:
I couldn't think of a better way to set the context specific for this thread within the task other than using a setter here. I was initially hoping that `execute()` could take something other than a runnable that way this enqueuing logic could construct some task type that added the context onto it and it got passed along to when it calls `httpClient.send()` but I couldn't figure out a way to do that.

Let me know if you can think of one.

Reply:
Actually, looking at `PriorityProcessWorkerExecutorService::executeWithPriority`. That's probably the better way to do it. Just throw in `execute()` and say it's not supported and use a custom method instead.